src/python/pants/backend/python/goals/publish.py

```diff
@@ -0,0 +1,198 @@
+    publish_request_type = PublishToPyPiRequest
+    required_fields = (PyPiRepositories,)
+
+    repositories: PyPiRepositories
+    skip_twine: SkipTwineUploadField
+
+    # I'd rather opt out early here, so we don't build unnecessarily, however the error feedback is
+    # misleading and not very helpful in that case.
+    #
+    # @classmethod
+    # def opt_out(cls, tgt: Target) -> bool:
+    #     return not tgt[PyPiRepositories].value
+
+
+def twine_upload_args(
+    twine_subsystem: TwineSubsystem, config_files: ConfigFiles, repo: str, dists: tuple[str, ...]
+) -> tuple[str, ...]:
+    args = ["upload", "--non-interactive"]
+
+    if config_files.snapshot.files:
+        args.append(f"--config-file={config_files.snapshot.files[0]}")
+
+    args.extend(twine_subsystem.args)
+
+    if repo.startswith("@"):
+        # Named repository from the config file.
+        args.append(f"--repository={repo[1:]}")
+    else:
+        args.append(f"--repository-url={repo}")
+
+    args.extend(dists)
+    return tuple(args)
+
+
+def twine_env_suffix(repo: str) -> str:
+    return f"_{repo[1:]}".replace("-", "_").upper() if repo.startswith("@") else ""
+
+
+def twine_env_request(repo: str) -> EnvironmentRequest:
+    suffix = twine_env_suffix(repo)
+    req = EnvironmentRequest(
+        [
+            f"{var}{suffix}"
+            for var in [
+                "TWINE_USERNAME",
+                "TWINE_PASSWORD",
+                "TWINE_REPOSITORY_URL",
+                # "TWINE_CERT",  # Does the --cert arg to pex take care of this for us?
+            ]
+        ]
+    )
+    return req
+
+
+def twine_env(env: Environment, repo: str) -> Environment:
+    suffix = twine_env_suffix(repo)
+    if not suffix:
+        return env
+
+    return Environment({key.rsplit(suffix, maxsplit=1)[0]: value for key, value in env.items()})
+
+
+@rule
+async def twine_upload(
+    request: PublishToPyPiRequest, twine_subsystem: TwineSubsystem
+) -> PublishPackagesProcesses:
+    dists = tuple(
+        artifact.relpath
+        for pkg in request.packages
+        for artifact in pkg.artifacts
+        if artifact.relpath
+    )
+
+    if twine_subsystem.skip or not dists:
+        return PublishPackagesProcesses(())
+
+    # Too verbose to provide feedback as to why some packages were skipped?
+    skip = None
+    if request.field_set.skip_twine.value:
+        skip = f"(by `{request.field_set.skip_twine.alias}` on {request.field_set.address})"
+    elif not request.field_set.repositories.value:
+        # I'd rather have used the opt_out mechanism on the field set, but that gives no hint as to
+        # why the target was not applicable..
+        skip = f"(no `{request.field_set.repositories.alias}` specifed for {request.field_set.address})"
+
+    if skip:
+        return PublishPackagesProcesses(
+            (
+                PublishPackageProcesses(
+                    names=dists,
+                    description=skip,
```

Comment:
It's helpful to see why some targets where skipped, when it wasn't intentional, but I'm worried if you use it on purpose, that it will be too chatty.
Perhaps hide it behind a verbose flag or something along those lines..?

Reply:
Seems fine for now I think... how likely is it that someone will be running `./pants publish ::` in a situation where there are local-only dists?

One general principle we try to apply though, is that 99% of warnings should have a way to opt out of the warning. In this case, this isn't actually rendered as a log warning, and so `--ignore-warnings` won't work. Another alternative would be to have a magical repository destination that specifies "local only", so if you have a dist that you only build locally without publishing, you could mark it to silence the warning.

Reply:
The log is a good switch. Rather than console output for those cases, then you may silence them.